accepted/00000-upgrading-bootstrap.md

```diff
@@ -0,0 +1,70 @@
+
+Bootstrap 3 comes from an era where browser support was more complicated and it reflects in the stylesheets we have. The current bundled stylesheet for Suma is roughly 10k lines of CSS, which is hard to reason about and maintain to say the least. This problem isn't solely due to Bootstrap, but it is the primary source of the issue. The manner in which Suma has been styled over the years has been essentially add only: since the styles are tightly coupled yet globally shared, removing things has been hard while adding a single override for any given problem is fairly easy. This has resulted in our own themes becoming complex as well, on top of an already large Bootstrap base.  
+
+As a separate issue, Bootstrap 3 has known security vulnerabilities, such as XSS injection, for which there won't be a patch. These security issues by themselves do not pose too serious of a threat due to the nature of the product, but they might be leveraged to create a chain of attacks. Given Suma is open source and has many moving parts, this is a very realistic attack vector.  
+
+The core proposition of this RFC can be summarized in two steps:  
+
+ 1. Simplify our existing stylesheets to reduce the use of custom overrides, and try and compose views and layouts using standard Bootstrap solutions. This step is beneficial even without the Bootstrap upgrade in mind, as currently our stylesheets are simply very convoluted and hard to reason about.  
+ 2. Upgrade Bootstrap. This includes many moving parts which will be outlined below.  
+
+Coming out successfully from the other side, we would hope to have simpler stylesheets with less to maintain, and fewer security issues.  
+
+
+# Detailed design
+[design]: #detailed-design
+
+## Theme & styles
+
+Bootstrap 3 uses Less while Bootstrap 5 only offers Sass styles. This switch by itself is not very problematic, however there are additional issues that make the upgrade more difficult. Our current styles make heave use of numerous mixins from Bootstrap 3 which no longer exist in Bootstrap 5. Additionally, there are numerous breaking styling changes going from 3 to 4 and even more so moving from 4 to 5.  
+
+Instead of trying to manually migrate our existing very large theme source, I propose we instead try to first reduce our existing theme footprint as much as possible. We already have component scoped styles working and in use, so we can try move component or view specific styles (e.g. table styles, setup wizard etc) into component scoped styles and remove them from the global stylesheet.  
+
+This allows us to cut up the effort required and do at least some of the work gradually. I suspect the last core theme change will still be painful, though. It might be easier to rebuild the existing theme from the base Bootstrap style rather than migrating the existing one, but this is something we can consider after we've simplified the theme.  
```

Comment:
Would the footprint reduction step be wasted if we ultimately decide to rewrite from scratch?

Reply:
Partially, there are still components for which this makes sense. Some examples would be our table component, the calendar widget we use, etc. But some of the effort would be wasted, yes. It's hard to guesstimate which would be less effort, as both detangling existing styles and styling Suma from scratch are both considerable challenges.

Reply:
I would not see it as a waste time, if we dive and learn better our code, by working in this reduction.

Reply:
> I would not see it as a waste time, if we dive and learn better our code, by working in this reduction.

Well, it would be a waste if all we learnt goes to the trash bin because we ultimately rewrite it all from scratch.

@Etheryte: the spirit of an RFC is to try to come up with the best possible alternative _given available evidence at the time of writing_. We can always course-correct later - just in principle (take it as an exercise) we strive to describe only one course of action as the suggested one by the RFC author, and all other alternatives in their own section.

In this case, I would encourage you to spend a bit of time and try to decide what path you see as most promising, so that only one proposal ends up in the main RFC text, and that will be our starting bet. As always we remain open to change plans in face of contrary evidence, and ready to switch bets!